### PR TITLE
ci: enable upload artifacts

### PR DESCRIPTION
**Pull request by v1v**
## What is the problem this PR solves?

Uses https://github.com/elastic/apm-pipeline-library/pull/1549 in order to upload the packages to a Google Storage bucket.

The folder structure is:

- `<bucket-name>/fleet-server/snapshots/`
- `<bucket-name>/fleet-server/pull-requests/pr-<id>/`
- `<bucket-name>/fleet-server/commit/<sha>/`

`<sha>` is the git commit related to the build and package happened in the CI
`pr-<id>` is the pull request ID

## Test

image

produced:

image

image

### CODE REVIEW

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -100,6 +102,7 @@ pipeline {
         dir("${BASE_DIR}"){
           withGoEnv(){
             sh(label: 'Create release artifacts', script: 'make docker-release')
+            uploadPackagesToGoogleBucket(pattern: 'build/binaries/**/*')
```
Comment (v1v) [inline .ci/Jenkinsfile:105]:
```suggestion
            uploadPackagesToGoogleBucket(pattern: 'build/binaries/*')
```

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -100,6 +102,7 @@ pipeline {
         dir("${BASE_DIR}"){
           withGoEnv(){
             sh(label: 'Create release artifacts', script: 'make docker-release')
+            uploadPackagesToGoogleBucket(pattern: 'build/binaries/*')
```
Comment (v1v) [inline .ci/Jenkinsfile:105]:
```suggestion
            uploadPackagesToGoogleBucket(pattern: 'build/binaries/**')
```

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env groovy
 
-@Library('apm@current') _
+@Library('apm@test/publish') _
```
Comment (v1v) [inline .ci/Jenkinsfile:3]: Requires a release

File: .ci/Jenkinsfile [hunk 4]
```diff
@@ -100,6 +102,7 @@ pipeline {
         dir("${BASE_DIR}"){
           withGoEnv(){
             sh(label: 'Create release artifacts', script: 'make docker-release')
+            uploadPackagesToGoogleBucket(pattern: 'build/binaries')
```
Comment (v1v) [inline .ci/Jenkinsfile:105]:
```suggestion
            uploadPackagesToGoogleBucket(pattern: 'build/binaries/*')
```

**Comment by v1v** [comment]
For some reason it copies files to a `binary` folder:

image

But it's not doing the same while using the `commits` folder:

image

File: .ci/Jenkinsfile (hunk 2 above)
Comment (v1v) [inline .ci/Jenkinsfile:105]:
```suggestion
            uploadPackagesToGoogleBucket(pattern: 'build/binaries/**/*')
```

File: .ci/Jenkinsfile (hunk 1 above)
Comment (v1v) [inline .ci/Jenkinsfile:105]:
```suggestion
            uploadPackagesToGoogleBucket(pattern: 'build/binaries/')
```

**Review by kuisathaverat — APPROVED** [review]

**Review by mdelapenya — COMMENTED** [review]
I've seen certain differences in how Beats and Fleet Server repo store the artifacts:

While Beats uses the following structure:

- 📁 beats
  - 📁 commits
    - 📁 SHA1
      - 📁 the_beat
        - 📦 [the_beat]-[version]-SNAPSHOT-[os]-[arch][extension], being [os] optional.

fleet-server uses the following structure:

- 📁 fleet-server
  - 📁 commits
    - 📁 SHA1
      - 📁: fleet-server-[version]-[os]-[arch], being [os] always present in the folder name.
        - 📦 fleet-server[extension]

Is that on purpose? I'd have expected having both follow the same structure, with Beats adding an intermediate dir for the_beat

**Comment by v1v** [comment]
> Is that on purpose? I'd have expected having both follow the same structure, with Beats adding an intermediate dir for the_beat


I'm a bit confused about this question, IIUC, the concern is regarding the intermediate folder, but fleet-server is not a monorepo, so there is no need to include an extra folder.

The fleet-server build system produces the output that's copied from, the only transformation done in this particular upload process within the CI is related to the `repo-name`/(`commits/sha`|`snapshots`|`pull-requests/pr-id`)

Is the issue related to the packaging system or the upload? if the later, what's the outcome?

**Comment by mdelapenya** [comment]
> Is the issue related to the packaging system or the upload? if the later, what's the outcome?

I believe it's a packaging issue. I'd have expected fleet-server to generate a child file:

- 📁 fleet-server
  - 📁 commits
    - 📁 SHA1
      - 📦 fleet-server-[version]-SNAPSHOT-[os]-[arch][extension], being [os] optional.

Furthermore, I realised that the ARM support is different: meanwhile Beats produces files including `aarch`, fleet-server does not:

Screenshot 2022-02-17 at 10 43 18
Screenshot 2022-02-17 at 10 43 33

**Comment by v1v** [comment]
> I believe it's a packaging issue. I'd have expected fleet-server to generate a child file including all variables:

What do we do?

> Furthermore, I realised that the ARM support is different: meanwhile Beats produces files including aarch, fleet-server does not:

Infra Release team pushed to use the same aarch extension from 8.0 onwards, but they asked to revert this change:

- https://github.com/elastic/beats/pull/28813#issuecomment-1024048002

We might need to support this :/

**Comment by mdelapenya** [comment]
Do you think it's something on the `mage package` command?

Screenshot 2022-02-17 at 15 57 42

Screenshot 2022-02-17 at 15 59 11

File: .ci/Jenkinsfile (hunk 3 above)
Comment (v1v) [reply .ci/Jenkinsfile:3]:
```suggestion
@Library('apm@current') _
```

**Comment by v1v** [comment]
### Acceptance Test


```
gsutil ls -r gs://beats-ci-artifacts/fleet-server/pull-requests/pr-1147 | grep '/fleet-server$'
gs://beats-ci-artifacts/fleet-server/pull-requests/pr-1147/fleet-server
gs://beats-ci-artifacts/fleet-server/pull-requests/pr-1147/fleet-server-8.2.0-darwin-arm64/fleet-server
gs://beats-ci-artifacts/fleet-server/pull-requests/pr-1147/fleet-server-8.2.0-darwin-x86_64/fleet-server
gs://beats-ci-artifacts/fleet-server/pull-requests/pr-1147/fleet-server-8.2.0-linux-arm64/fleet-server
gs://beats-ci-artifacts/fleet-server/pull-requests/pr-1147/fleet-server-8.2.0-linux-x86/fleet-server
gs://beats-ci-artifacts/fleet-server/pull-requests/pr-1147/fleet-server-8.2.0-linux-x86_64/fleet-server
```

```
gsutil ls -r gs://beats-ci-artifacts/fleet-server/commits/b19f7c0713f6991c7f6e6ad5cdefad78fe5c37de | grep '/fleet-server$'
gs://beats-ci-artifacts/fleet-server/commits/b19f7c0713f6991c7f6e6ad5cdefad78fe5c37de/fleet-server-8.2.0-darwin-arm64/fleet-server
gs://beats-ci-artifacts/fleet-server/commits/b19f7c0713f6991c7f6e6ad5cdefad78fe5c37de/fleet-server-8.2.0-darwin-x86_64/fleet-server
gs://beats-ci-artifacts/fleet-server/commits/b19f7c0713f6991c7f6e6ad5cdefad78fe5c37de/fleet-server-8.2.0-linux-arm64/fleet-server
gs://beats-ci-artifacts/fleet-server/commits/b19f7c0713f6991c7f6e6ad5cdefad78fe5c37de/fleet-server-8.2.0-linux-x86/fleet-server
gs://beats-ci-artifacts/fleet-server/commits/b19f7c0713f6991c7f6e6ad5cdefad78fe5c37de/fleet-server-8.2.0-linux-x86_64/fleet-server
```

**Review by mdelapenya — APPROVED** [review]
With latest changes in https://github.com/elastic/e2e-testing/pull/2172, I think it would be easy to create an edge case for Fleet Server

**Comment by v1v** [comment]
I'll merge this now

**Pull request merged**